Notify user on broken PDF files

dcaliste opened the pull request:
Proposed patches to avoid crashes when tapping on a broken PDF file (issue #21).

I've :
- protected all calls to attributes of Poppler::Document when it is a null pointer because of loading failure at pdfjob.cpp:31.
- added a flag in PdfRenderThread that becomes true when file loading results in a null pointer and expose it read only in PdfDocument.
- used this flag in QML to return to document list when tapping on a broken PDF file and notify the user.

What do you think ? I don't like much the fact that the page for the PDF file is pushed and then popped immediately, but I don't know what to show otherwise. On the other side, that would be nice to still have the drawer at hand even for broken PDF files, because we may want to forward them by email for investigation…

VDVsx commented on pdf/pdfrenderthread.cpp:
```diff
@@ -143,7 +144,7 @@ PDFRenderThread::~PDFRenderThread()
 {
     QMutexLocker locker{ &d->thread->mutex };
-    return d->document->numPages();
+    return (d->document != nullptr)?d->document->numPages():0;
```
Add extra spaces here between ? and : conditions for better readability.

VDVsx commented on pdf/pdfrenderthread.cpp:
```diff
@@ -214,8 +220,14 @@ void PDFRenderThreadQueue::processPendingJob()
+        }
+        default:
+            job->m_document = d->document;
+    }
```
This project is probably not following the Qt conventions in other places, but we aim follow them, see here: http://qt-project.org/wiki/Qt_Coding_Style#e289ee44592e9c32d4212069f0806daf

VDVsx commented on plugin/PDFDocumentPage.qml:
```diff
@@ -55,4 +61,11 @@ DocumentPage {
+        category: "x-sailfish.sailfish-office.error";
+        previewBody: "Cannot open PDF file";
+        previewSummary: "Broken PDF file";
+    }    
```
This need to be translatable:

//% "Cannot open PDF file"
previewBody:qsTrId(... ;

check other translations in this project to match the same naming.

VDVsx commented on rpm/sailfish-office.spec:
```diff
@@ -44,6 +44,7 @@ Group: System/Base
 %{_datadir}/dbus-1/interfaces/
 %{_datadir}/dbus-1/services/
+%{_datadir}/lipstick/notificationcategories/x-sailfish.sailfish-office.error.conf
 
```
For in app notifications this is not needed, we don't want the notification to show up in events view.

VDVsx commented on plugin/PDFDocumentPage.qml:
```diff
@@ -44,6 +45,11 @@ DocumentPage {
+        onDocumentLoaded: if (failure) {
+            pageStack.completeAnimation();
+            pageStack.pop();
+            notification.publish();
```
This part I need to test on device to see if behavior can be improved, will take a bit more.

dcaliste replied:
Yep, that's great of you. Currently the page is pushed and then the PDF document is loaded, so I don't know how not to push the page without being too intrusive in the code.

Besides, this may be irrelevant not to push the page because, I'm preparing a kind of "place holder" for PDF without rendering (because of broken file or because the file is valid but currently locked down). This place holder would replace the PDFView Item and display a message that the PDF is either broken or protected. In case of protected file a button would open a dialog with a text input to enter the password (related to some discussion in issue #1). Having this place holder would keep the actions of the drawer available, like sharing the file by bluetooth…

What do you think ?

VDVsx replied:
Sounds good since it covers additional cases.

dcaliste commented:
I've taken into account your remarks and rebased my two commits hopefully in a better shape :
* Removed the x-sailfish.sailfish-office.error.conf file ;
* Try to follow the Qt coding rules for switches and add spaces for better readability ;
* Use translatable strings for the notification.

Beside, how should I handle the translations ? At least in English ?

VDVsx commented:
You just need the engineering English like is done now, Jolla translation team does the rest.

VDVsx commented:
I'm merging this already to facilitate your work, then can tag later after next commit that addresses the QML side like discussed above.

dcaliste commented:
Thank you.
